contracts/cw-names-registry/src/contract.rs

```diff
@@ -0,0 +1,330 @@
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(wrapped) => execute_receive(deps, env, info, wrapped),
+        ExecuteMsg::RegisterName { name } => execute_register_name_native(deps, env, info, name),
+        ExecuteMsg::UpdateConfig {
+            new_admin,
+            new_payment_info,
+        } => execute_update_config(deps, env, info, new_admin, new_payment_info),
+        ExecuteMsg::Revoke { name } => execute_revoke(deps, env, info, name),
+        ExecuteMsg::Reserve { name } => execute_reserve(deps, env, info, name),
+        ExecuteMsg::TransferReservation { name, dao } => {
+            execute_transfer_reservation(deps, env, info, name, dao)
+        }
+    }
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    wrapped: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    match config.payment_info {
+        PaymentInfo::NativePayment { .. } => Err(ContractError::Unauthorized {}),
+        PaymentInfo::Cw20Payment {
+            token_address,
+            payment_amount,
+        } => {
+            if info.sender != token_address {
+                return Err(ContractError::UnrecognisedCw20 {});
+            }
+
+            let sender = wrapped.sender;
+            let amount = wrapped.amount;
+            let msg: ReceiveMsg = from_binary(&wrapped.msg)?;
+
+            match msg {
+                ReceiveMsg::Register { name } => {
+                    register_name(deps, env, sender, amount, name, payment_amount)
+                }
+            }
+        }
+    }
+}
+
+pub fn execute_register_name_native(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    name: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    match config.payment_info {
+        PaymentInfo::Cw20Payment { .. } => Err(ContractError::InvalidPayment {}),
+        PaymentInfo::NativePayment {
+            token_denom,
+            payment_amount,
+        } => {
+            let token_idx = info.funds.iter().position(|c| c.denom == token_denom);
+            if token_idx.is_none() {
+                return Err(ContractError::UnrecognisedNativeToken {});
+            }
+
+            let coins = &info.funds[token_idx.unwrap()];
+
+            register_name(
+                deps,
+                env,
+                info.sender.to_string(),
+                coins.amount,
+                name,
+                payment_amount,
+            )
+        }
+    }
+}
+
+pub fn register_name(
+    deps: DepsMut,
+    _env: Env,
+    sender: String,
+    amount_sent: Uint128,
+    name: String,
+    payment_amount_to_register_name: Uint128,
+) -> Result<Response, ContractError> {
+    if amount_sent != payment_amount_to_register_name {
+        return Err(ContractError::InsufficientFunds {});
```

Comment:
Is `InsufficientFunds` the right error here? It looks like we're also erroring if the funds are too high. I can see this confusing someone.